turtle_tf2_cpp/src/turtle_tf2_listener.cpp

```diff
@@ -88,42 +66,92 @@ class FrameListener : public rclcpp::Node
     std::string fromFrameRel = target_frame_.c_str();
     std::string toFrameRel = "turtle2";
 
-    geometry_msgs::msg::TransformStamped transformStamped;
-
-    // Look up for the transformation between target_frame and turtle2 frames
-    // and send velocity commands for turtle2 to reach target_frame
-    try {
-      transformStamped = tf_buffer_->lookupTransform(
-        toFrameRel, fromFrameRel,
-        tf2::TimePoint(),
-        500ms);
-    } catch (tf2::LookupException & ex) {
-      RCLCPP_INFO(this->get_logger(), "transform not ready");
-      return;
+    if (!turtle_spawned_) {
+      // Check if the service is available
+      while (!spawner_->wait_for_service(1s)) {
+        if (!rclcpp::ok()) {
+          RCLCPP_ERROR(
+            this->get_logger(),
+            "Interrupted while waiting for the service. Exiting."
+          );
+          timer_->cancel();
+          return;
+        }
+        RCLCPP_INFO(
+          this->get_logger(),
+          "Service not available, waiting again..."
+        );
+      }
+
+      // Initialize request with turtle name and coordinates
+      // Note that x, y and theta are defined as floats in turtlesim/srv/Spawn
+      auto request = std::make_shared<turtlesim::srv::Spawn::Request>();
+      request->x = 4.0;
+      request->y = 2.0;
+      request->theta = 0.0;
+      request->name = "turtle2";
+      // Call request
+      auto result = spawner_->async_send_request(request);
+
+      turtle_spawned_ = true;
     }
 
-    geometry_msgs::msg::Twist msg;
+    if (!transform_available_) {
+      const auto timeout_ms_ = 100ms;
+      while (!tf_buffer_->canTransform(toFrameRel, fromFrameRel, tf2::TimePointZero, 0ms) &&
+        rclcpp::ok())
+      {
+        RCLCPP_INFO(
+          get_logger(), "Waiting %ld ms for %s->%s transform to become available",
+          timeout_ms_.count(), toFrameRel.c_str(), fromFrameRel.c_str());
+        std::this_thread::sleep_for(timeout_ms_);
+      }
+      RCLCPP_INFO(
+        get_logger(), "Transform %s->%s available",
+        toFrameRel.c_str(), fromFrameRel.c_str());
+      transform_available_ = true;
+    }
```

Comment:
Again, we don't want to sleep here waiting for the transform.  That holds up everything else in the executor.  Here again I'd call `canTransform` with a 0ms wait, and if that fails just return immediately.  Subsequent `on_timer` calls will try again.

(alternatively, you could get rid of the call to `canTransform` and just rely on `lookupTransform` to do that for you below, but that depends on what you are trying to teach here)

Reply:
Thanks! That made code neater. Solved in https://github.com/ros/geometry_tutorials/pull/53/commits/4a380e7cc5b293228d224a429d4d5fbce7a168db

Reply:
I removed the part with canTransform() to make code more readable https://github.com/ros/geometry_tutorials/pull/53/commits/07e71cff765e63b4423bd4e0cf70e9e456e3ae20